packages/api/server/ws.mts

```diff
@@ -306,16 +282,109 @@ async function cellUpdate(payload: CellUpdatePayloadType) {
     wss.broadcast(`session:${session.id}`, 'cell:updated', {
       cell: findCell(session, payload.cellId),
     });
+
+    return;
+  }
+
+  if (
+    session.metadata.language === 'typescript' &&
+    result.cell.type === 'code' &&
+    tsservers.has(session.id)
+  ) {
+    const cell = result.cell;
+    const tsserver = tsservers.get(session.id);
+
+    const file = Path.join(session.dir, cell.filename);
+
+    // To update a file in tsserver, close and reopen it. I assume performance of
+    // this implementation is worse than calculating diffs and using `change` command
+    // (although maybe not since this is not actually reading or writing to disk).
+    // However, that requires calculating diffs which is more complex and may also
+    // have performance implications, so sticking with the simple approach for now.
+    //
+    // TODO: In either case, we should be aggressively debouncing these updates.
+    //
+    tsserver.close({ file });
+    tsserver.open({ file, fileContent: cell.source });
+
+    sendTypeScriptDiagnostics(tsserver, session, cell);
   }
 }
 
+/**
+ * Send semantic diagnostics for a TypeScript cell to the client.
+ */
+async function sendTypeScriptDiagnostics(
+  tsserver: TsServer,
+  session: SessionType,
+  cell: CodeCellType,
+) {
+  const response = await tsserver.semanticDiagnosticsSync({
+    file: Path.join(session.dir, cell.filename),
+  });
+
+  const diagnostics = response.body || [];
+
+  for (const diagnostic of diagnostics) {
+    wss.broadcast(`session:${session.id}`, 'cell:output', {
+      cellId: cell.id,
+      output: {
+        type: 'tsc',
+        data: formatDiagnostic(diagnostic),
+      },
+    });
+  }
+}
+
+function sendAllTypeScriptDiagnostics(tsserver: TsServer, session: SessionType) {
+  for (const cell of session.cells) {
+    if (cell.type === 'code') {
+      sendTypeScriptDiagnostics(tsserver, session, cell);
+    }
+  }
+}
+
+async function tsserverStart(payload: TsServerStartPayloadType) {
+  const session = await findSession(payload.sessionId);
+
+  if (!session) {
+    throw new Error(`No session exists for session '${payload.sessionId}'`);
+  }
+
+  if (session.metadata.language !== 'typescript') {
+    throw new Error(`tsserver can only be used with TypeScript Srcbooks.`);
+  }
+
+  if (!tsservers.has(session.id)) {
```

Comment:
Should it restart if it exists? Do you think we'd ever call this to kick it?

Reply:
I don't think implicitly writing code to account for a potential future problem is a good idea. Either do it explicitly when the problem arises or not at all

Reply:
Restarting is also inefficient compared to doing nothing if it's already running